### ISSUE TITLE
ClosedXML - Workbook.saveas() giving ArguementOutOfRange exception

### ISSUE THREAD

**Issue by atulverma84** [ClosedXML/ClosedXML#994]
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

If you are logging a possible bug or feature request, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**

Latest version

**What is the current behavior?**

I have a workbook with 8 tables with formulas. I want to copy the same worksheet 'n' number of times with tables. when i call the worksheet.saveas(memorystream) , it gives the following error 👎 

Specified argument was out of the range of valid values.  
at DocumentFormat.OpenXml.Packaging.OpenXmlPartContainer.GetPartById(String id)
   at ClosedXML.Excel.XLWorkbook.GenerateTables(XLWorksheet worksheet, WorksheetPart worksheetPart, SaveContext context, XLWSContentManager cm)
   at ClosedXML.Excel.XLWorkbook.GenerateWorksheetPartContent(WorksheetPart worksheetPart, XLWorksheet xlWorksheet, Boolean evaluateFormulae, SaveContext context)
   at ClosedXML.Excel.XLWorkbook.CreateParts(SpreadsheetDocument document, SaveOptions options)
   at ClosedXML.Excel.XLWorkbook.CreatePackage(Stream stream, Boolean newStream, SpreadsheetDocumentType spreadsheetDocumentType, SaveOptions options)
   at ClosedXML.Excel.XLWorkbook.SaveAs(Stream stream, SaveOptions options)
   at Accenture.Generics.DataModels.Reports.QueryHandlerExcel.HandleQuery(ReportDataRequest dataRequest, Dictionary`2 queryStrings, Int32 timeout)

**What is the expected behavior or new feature?**

Complete this.

**Did this work in previous versions of our tool?  Which versions?**

Yes/No/v0.XX

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
using (XLWorkbook workBook = new XLWorkbook(excelTemplate))
                    {
                            // execute all data query in the definition
                            foreach (var item in dataRequest.DataRequests)
                            {
                                string tableName = item.Key;
                                var dataQuery = item.Value;
                                if (item.Value.NestedDataRequests.Count > 0)
                                {
                                    if (item.Value.HandlerType == "SQL")
                                    {
                                        QueryHandlerSQL queryHandlerSql = new QueryHandlerSQL();
                                        resultdata = queryHandlerSql.GetSqlData(dataQuery, queryStrings, timeout);
                                    }
                                    else if (item.Value.HandlerType == "MDX")
                                    {
                                        QueryHandlerMDX queryHandlerMdx = new QueryHandlerMDX();
                                        resultdata = queryHandlerMdx.GetMdxData(dataQuery, queryStrings, timeout);
                                    }
                                }

                                DataTable dt = GetDataTable(resultdata, item.Value.HandlerType, item.Value);
                                bool isFirstSheet = true;
                                bool isNotFirstSheet = false;
                                IXLWorksheet dataSheet = null;
                                IXLWorksheet firstWorksheet = null;
                                IXLWorksheet worksheet = null;
                               //// workBook.Worksheet(1).CopyTo("abc");
                                ////int i = 0;
                                foreach (DataRow dr in dt.Rows)
                                {
                                    workBook.Worksheet(1).CopyTo(dr.ItemArray[1].ToString());
                                }

                                ////workBook.Worksheet(1).Delete();
                                
                               foreach (DataRow dr in dt.Rows)
                                {
                                    if (isFirstSheet)
                                    {
                                        dataSheet = workBook.Worksheet(dr.ItemArray[1].ToString());
                                        isFirstSheet = false;
                                    }
                                    else
                                    {
                                       //// firstWorksheet = workBook.Worksheet(1).CopyTo("FirstSheet");
                                        isNotFirstSheet = true;
                                        dataSheet = null;
                                        dataSheet = workBook.Worksheet(dr.ItemArray[1].ToString());
                                        isFirstSheet = false;
                                    }

                                    dataSheet.Name = dataSheet.RangeAddress.Worksheet.Name = dr.ItemArray[1].ToString();

                                    foreach (var multiItem in item.Value.NestedDataRequests)
                                    {
                                        foreach (var request in multiItem.Values)
                                        {
                                            if (!queryStrings.ContainsKey("BuyingAudienceId"))
                                            {
                                                queryStrings.Add("BuyingAudienceId", dr.ItemArray[0].ToString());
                                            }
                                            else
                                            {
                                                queryStrings["BuyingAudienceId"] = dr.ItemArray[0].ToString();
                                            }

                                            var getdata = this.GetData(request, queryStrings, timeout);

                                            // Retrieve data for data Query definition
                                            ReportData data = new ReportData();

                                            if (getdata == "[]")
                                            {
                                                ////fill excel table defined by TableName
                                                //// worksheet = this.FillCreateWorksheet(dataSheet, data.Data(), dr.ItemArray[1].ToString(), null, multiItem.Keys.FirstOrDefault(), dr.ItemArray[0].ToString(), request.InsertRows, isFirstSheet, isNotFirstSheet, request.ShowTotal);
                                                break;
                                            }

                                            data.ExportData = getdata;

                                            ////fill excel table defined by TableName
                                             worksheet = this.FillCreateWorksheet(dataSheet, data.Data(), dr.ItemArray[1].ToString(), null, multiItem.Keys.FirstOrDefault(), dr.ItemArray[0].ToString(), request.InsertRows, isFirstSheet,  isNotFirstSheet, request.ShowTotal);

                                            if (request.Actions != null && request.Actions.Count != 0)
                                            {
                                                QueryHandlerExcel.ExecuteActions(request.Actions, workBook);
                                            }                                       
                                            }                           
                                        }
                                     }
                                  }

                            // store file in memmory and return byte array
                            using (MemoryStream resultStream = new MemoryStream())
                            {
                                workBook.SaveAs(resultStream);
                                returnArray = DynamicTypeBuilder.Serialize(resultStream.ToArray());
                                resultStream.Flush();
                            }
                        }
                    }

public IXLWorksheet FillCreateWorksheet(IXLWorksheet workSheet, Dictionary<string, object>[] data, string reportName, string[] headers, string tableName, string audienceId, bool rowsAdded, bool isfirstSheet, bool isNotFirstSheet, bool showTotal)
        {
            try
            {
                if (workSheet == null)
                {
                    throw new ArgumentNullException("workSheet");
                }

                if (data == null)
                {
                    throw new ArgumentNullException("data");
                }
                
                ////workSheet.Cell(1,1).Value = "";
                if (reportName == null)
                {
                    throw new ArgumentNullException("reportName");
                }

                // define header
                if (headers == null)
                {
                    if (data.Count() > 0)
                    {
                        headers = data[0].Keys.ToArray();
                    }
                    else
                    {
                        headers = new string[0];
                    }
                }

                using (TimingTrace target = new TimingTrace(this.controllerName, "AddWorksheet"))
                {                    
                    var aud = audienceId;

                    workSheet.Cell("A4").Value = reportName;
                    
                    IXLTable table = workSheet.Table(tableName);

                    // Create Header in the worksheet
                    IXLCell lastCell = table.LastCell();
                    IXLCell firstCell = table.FirstCell();

                    int rowPosition = firstCell.WorksheetRow().RowNumber();
                    int columnPosion = firstCell.WorksheetColumn().ColumnNumber();

                    int lastColumnPostion = lastCell.WorksheetColumn().ColumnNumber();

                    foreach (var column in headers)
                    {
                        workSheet.Row(rowPosition).Cell(columnPosion).Value = column;
                        columnPosion++;
                    }

                    //////// Check data values
                    //////if (data.Count() == 0)
                    //////{
                    //////    workSheet.Visibility = XLWorksheetVisibility.VeryHidden;
                    //////}

                    if (rowsAdded && !isNotFirstSheet)
                    {
                        var s = isfirstSheet;
                        workSheet.Row(firstCell.Address.RowNumber + 2).InsertRowsBelow(data.Count());                        
                    }
                    
                    // add data 
                    foreach (var row in data)
                    {
                        rowPosition++;
                        columnPosion = firstCell.WorksheetColumn().ColumnNumber();
                        foreach (var column in headers)
                        {
                            var value = row[column];
                            if (value != null)
                            {
                                if (value is string)
                                {
                                    workSheet.Row(rowPosition).Cell(columnPosion).Value = value.ToString();
                                }
                                else
                                {
                                    if (!value.GetType().IsClass)
                                    {
                                        workSheet.Row(rowPosition).Cell(columnPosion).Value = value;
                                    }
                                }
                            }

                            lastCell = workSheet.Row(rowPosition).Cell(columnPosion);
                            columnPosion++;
                        }
                    }

                    if (showTotal && isNotFirstSheet && data.Count() > 0)
                    {
                        lastCell = workSheet.Row(rowPosition + 1).Cell(columnPosion - 1);
                    }

                  workSheet.Tables.Remove(tableName);
                   workSheet.Range(firstCell, lastCell).CreateTable(tableName);

                    if (showTotal)
                    {
                        table = workSheet.Table(tableName);
                        table.ShowTotalsRow = true;
                        foreach (IXLTableField x in table.Fields)
                        {
                            x.TotalsRowFunction = XLTotalsRowFunction.Sum;
                        }

                        table.Field(0).TotalsRowLabel = "Total";
                    }  

                    return workSheet;
                }
            }
            catch (Exception ex)
            {
                throw ExceptionManager.LogException(ex, null, string.Empty);
            }
        }

```
[Ireland_Multiple_table_Template - Copy.xlsx](https://github.com/ClosedXML/ClosedXML/files/2373441/Ireland_Multiple_table_Template.-.Copy.xlsx)

**Comment by vbjay** [comment]
https://stackoverflow.com/help/mcve

**Comment by atulverma84** [comment]
I tried the same too.. when i remove following lines

   workSheet.Tables.Remove(tableName);
   workSheet.Range(firstCell, lastCell).CreateTable(tableName);

saveas() is working fine. But it is not expanding my tables in resulting worksheets.

**Comment by vbjay** [comment]
The details are helpful.  However, you failed in providing an MVCE.  An mvce is something we could copy into a console program and maybe alter  path to the excel file and run.  We cannot run your code.

**Comment by atulverma84** [comment]
Thanks for quick response.

Here is code to run in console application -

attached is the sample file.

 static void Main(string[] args)
        {

                using (XLWorkbook workBook = new XLWorkbook("C:\\report.xlsx"))
                {

                    IXLWorksheet xl = workBook.Worksheet(1);
                    xl.CopyTo(workBook, "Abc");

                IXLWorksheet w = workBook.Worksheet("Abc");
                IXLCell lastCell = w.Table("DayPart").LastCell();
                IXLCell firstCell = w.Table("DayPart").FirstCell();
                w.Tables.Remove("DayPart");
                w.Range(firstCell, lastCell).CreateTable("DayPart");
           
                workBook.Save();
                }

        }
[report.xlsx](https://github.com/ClosedXML/ClosedXML/files/2373629/report.xlsx)

On save it is giving following error 👎 

System.ArgumentOutOfRangeException
  HResult=0x80131502
  Message=Specified argument was out of the range of valid values.
Parameter name: id
  Source=DocumentFormat.OpenXml
  StackTrace:
   at DocumentFormat.OpenXml.Packaging.OpenXmlPartContainer.GetPartById(String id)
   at ClosedXML.Excel.XLWorkbook.GenerateTables(XLWorksheet worksheet, WorksheetPart worksheetPart, SaveContext context, XLWSContentManager cm)
   at ClosedXML.Excel.XLWorkbook.GenerateWorksheetPartContent(WorksheetPart worksheetPart, XLWorksheet xlWorksheet, Boolean evaluateFormulae, SaveContext context)
   at ClosedXML.Excel.XLWorkbook.CreateParts(SpreadsheetDocument document, SaveOptions options)
   at ClosedXML.Excel.XLWorkbook.CreatePackage(String filePath, SpreadsheetDocumentType spreadsheetDocumentType, SaveOptions options)
   at ClosedXML.Excel.XLWorkbook.Save(SaveOptions options)
   at ClosedXML.Excel.XLWorkbook.Save(Boolean validate, Boolean evaluateFormulae)
   at ClosedXML.Excel.XLWorkbook.Save()
   at ConsoleApp1.Program.Main(String[] args) in

**Comment by atulverma84** [comment]
Any update on above issue ? are you able to run the application in console. ?

**Comment by igitur** [comment]
I could reproduce the error and fixed it in #998
